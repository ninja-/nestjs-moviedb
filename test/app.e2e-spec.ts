import * as request from 'supertest';
import { Test } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import { MovieDatabaseService } from '../src/movie-database.service';
import { MovieController } from '../src/movie.controller';
import { MovieRandomizerService } from '../src/movie-randomizer.service';

// Usually I'd prepare a sepeare layer of tests on controller level but it's too much of a hassle for now.

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleFixture = await Test.createTestingModule({
      providers: [
        MovieDatabaseService, 
        MovieRandomizerService,
        {
          provide: 'DB_NO_SAVE',
          useValue: true
        },
        {
          provide: 'DB_FILE',
          useValue: 'data/db-test.json'
        }    
      ],
      controllers: [MovieController]
    }).compile();

    await moduleFixture.init()

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        transform: true,
        whitelist: true
      }),
    );
      
    await app.init();
  });

  it('Properly inserts new movie (PUT)', async () => {
    // Edge cases for future tests:
    // - exceeding allowed field length
    // - invalid genres in array
    // - not providing required fields

    let movie = {
      title: 'Test movie',
      director: 'Tim',
      genres: ["Comedy"],
      plot: "bla",
      posterUrl: "",
      runtime: 92,
      year: 2020
    }

    const response = await request(app.getHttpServer())
      .put('/movies/')
      .send({movie: movie})
    expect(response.status).toEqual(200)
    expect(response.body).toEqual(movie)

    // Find it on global list now
    const allMovies = await request(app.getHttpServer())
      .get('/movies/')
    const myMovie = allMovies.body.find(m => m.title == movie.title)
    expect(myMovie).toEqual(movie)
  });
  
  it('Properly chooses random movies based on algorithm', async () => {
    // Edge cases for future tests:
    // - handling limit
    // - empty genres[]

    const args = {
      limit: 10,
      genres: ["Comedy", "Drama"],
      duration: 110
    }

    // Store a snapshot
    // I am a big fan of snapshot testing lately :D
    // That comes from experience in a project where changing say user model broke 50 different tests because of hardcoded data :(
    // And with snapshots this could be automated by using jest --updateSnapshots

    const response = await request(app.getHttpServer())
      .get('/movies/random')
      .send(args)

    const body = response.body
    expect(body).toMatchSnapshot()

  })
});
