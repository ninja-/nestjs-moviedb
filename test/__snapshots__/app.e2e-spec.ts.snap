// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) Properly chooses random movies based on algorithm 1`] = `
Array [
  Object {
    "actors": "François Cluzet, Omar Sy, Anne Le Ny, Audrey Fleurot",
    "director": "Olivier Nakache, Eric Toledano",
    "genres": Array [
      "Biography",
      "Comedy",
      "Drama",
    ],
    "plot": "After he becomes a quadriplegic from a paragliding accident, an aristocrat hires a young man from the projects to be his caregiver.",
    "posterUrl": "http://ia.media-imdb.com/images/M/MV5BMTYxNDA3MDQwNl5BMl5BanBnXkFtZTcwNTU4Mzc1Nw@@._V1_SX300.jpg",
    "runtime": 112,
    "title": "The Intouchables",
    "year": 2011,
  },
  Object {
    "actors": "Ethan Suplee, Ben Affleck, Scott Mosier, Jason Lee",
    "director": "Kevin Smith",
    "genres": Array [
      "Comedy",
      "Drama",
      "Romance",
    ],
    "plot": "Holden and Banky are comic book artists. Everything's going good for them until they meet Alyssa, also a comic book artist. Holden falls for her, but his hopes are crushed when he finds out she's gay.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BZDM3MTg2MGUtZDM0MC00NzMwLWE5NjItOWFjNjA2M2I4YzgxXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
    "runtime": 113,
    "title": "Chasing Amy",
    "year": 1997,
  },
  Object {
    "actors": "Jim Carrey, Laura Linney, Noah Emmerich, Natascha McElhone",
    "director": "Peter Weir",
    "genres": Array [
      "Comedy",
      "Drama",
      "Sci-Fi",
    ],
    "plot": "An insurance salesman/adjuster discovers his entire life is actually a television show.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMDIzODcyY2EtMmY2MC00ZWVlLTgwMzAtMjQwOWUyNmJjNTYyXkEyXkFqcGdeQXVyNDk3NzU2MTQ@._V1_SX300.jpg",
    "runtime": 103,
    "title": "The Truman Show",
    "year": 1998,
  },
  Object {
    "actors": "Brendan Gleeson, Chris O'Dowd, Kelly Reilly, Aidan Gillen",
    "director": "John Michael McDonagh",
    "genres": Array [
      "Comedy",
      "Drama",
    ],
    "plot": "After he is threatened during a confession, a good-natured priest must battle the dark forces closing in around him.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTc3MjQ1MjE2M15BMl5BanBnXkFtZTgwNTMzNjE4MTE@._V1_SX300.jpg",
    "runtime": 102,
    "title": "Calvary",
    "year": 2014,
  },
  Object {
    "actors": "Michael Keaton, Emma Stone, Kenny Chin, Jamahl Garrison-Lowe",
    "director": "Alejandro G. Iñárritu",
    "genres": Array [
      "Comedy",
      "Drama",
      "Romance",
    ],
    "plot": "Illustrated upon the progress of his latest Broadway play, a former popular actor's struggle to cope with his current life as a wasted actor is shown.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BODAzNDMxMzAxOV5BMl5BanBnXkFtZTgwMDMxMjA4MjE@._V1_SX300.jpg",
    "runtime": 119,
    "title": "Birdman or (The Unexpected Virtue of Ignorance)",
    "year": 2014,
  },
  Object {
    "actors": "Bruce Dern, Will Forte, June Squibb, Bob Odenkirk",
    "director": "Alexander Payne",
    "genres": Array [
      "Adventure",
      "Comedy",
      "Drama",
    ],
    "plot": "An aging, booze-addled father makes the trip from Montana to Nebraska with his estranged son in order to claim a million-dollar Mega Sweepstakes Marketing prize.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTU2Mjk2NDkyMl5BMl5BanBnXkFtZTgwNTk0NzcyMDE@._V1_SX300.jpg",
    "runtime": 115,
    "title": "Nebraska",
    "year": 2013,
  },
  Object {
    "actors": "Patton Oswalt, Ian Holm, Lou Romano, Brian Dennehy",
    "director": "Brad Bird, Jan Pinkava",
    "genres": Array [
      "Animation",
      "Comedy",
      "Family",
    ],
    "plot": "A rat who can cook makes an unusual alliance with a young kitchen worker at a famous restaurant.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTMzODU0NTkxMF5BMl5BanBnXkFtZTcwMjQ4MzMzMw@@._V1_SX300.jpg",
    "runtime": 111,
    "title": "Ratatouille",
    "year": 2007,
  },
  Object {
    "actors": "Benicio Del Toro, Dennis Farina, Vinnie Jones, Brad Pitt",
    "director": "Guy Ritchie",
    "genres": Array [
      "Comedy",
      "Crime",
    ],
    "plot": "Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers, and supposedly Jewish jewelers fight to track down a priceless stolen diamond.",
    "posterUrl": "http://ia.media-imdb.com/images/M/MV5BMTA2NDYxOGYtYjU1Mi00Y2QzLTgxMTQtMWI1MGI0ZGQ5MmU4XkEyXkFqcGdeQXVyNDk3NzU2MTQ@._V1_SX300.jpg",
    "runtime": 102,
    "title": "Snatch",
    "year": 2000,
  },
  Object {
    "actors": "Jason Flemyng, Dexter Fletcher, Nick Moran, Jason Statham",
    "director": "Guy Ritchie",
    "genres": Array [
      "Comedy",
      "Crime",
    ],
    "plot": "A botched card game in London triggers four friends, thugs, weed-growers, hard gangsters, loan sharks and debt collectors to collide with each other in a series of unexpected events, all for the sake of weed, cash and two antique shotguns.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTAyN2JmZmEtNjAyMy00NzYwLThmY2MtYWQ3OGNhNjExMmM4XkEyXkFqcGdeQXVyNDk3NzU2MTQ@._V1_SX300.jpg",
    "runtime": 107,
    "title": "Lock, Stock and Two Smoking Barrels",
    "year": 1998,
  },
  Object {
    "actors": "Joel Murray, Tara Lynne Barr, Melinda Page Hamilton, Mackenzie Brooke Smith",
    "director": "Bobcat Goldthwait",
    "genres": Array [
      "Comedy",
      "Crime",
    ],
    "plot": "On a mission to rid society of its most repellent citizens, terminally ill Frank makes an unlikely accomplice in 16-year-old Roxy.",
    "posterUrl": "https://images-na.ssl-images-amazon.com/images/M/MV5BMTQwMTc1MzA4NF5BMl5BanBnXkFtZTcwNzQwMTgzNw@@._V1_SX300.jpg",
    "runtime": 105,
    "title": "God Bless America",
    "year": 2011,
  },
]
`;
